apps/scratch/src/pages/__tests__/page-objects/scratch-editor.ts

```diff
@@ -92,4 +103,79 @@ export class ScratchEditorPage {
       await this.page.waitForTimeout(1000);
     }
   }
+
+  async scrollBlockIntoView(scrollToBlock: Locator) {
+    const toolboxMiddle = await this.toolbox.evaluate(
+      (toolbox) =>
+        toolbox.getBoundingClientRect().top +
+        toolbox.getBoundingClientRect().height / 2,
+    );
+
+    const toolboxRight = await this.toolbox.evaluate(
+      (toolbox) =>
+        toolbox.getBoundingClientRect().x +
+        toolbox.getBoundingClientRect().width,
+    );
+
+    await this.page.mouse.move(toolboxRight + 100, toolboxMiddle);
+
+    const offsetX =
+      (await scrollToBlock.evaluate(
+        (canvas) => canvas.getBoundingClientRect().x,
+      )) - toolboxRight;
+
+    // scroll all the way to the right so that
+    await this.page.mouse.wheel(offsetX, 0);
+  }
+
+  removeBlock(block: Locator) {
+    return block.dragTo(this.toolbox, { force: true });
+  }
+
+  async appendNewBlockTo(opcode: string, block: Locator) {
+    await this.scrollBlockIntoView(block);
+
+    await this.getBlockInToolbox(opcode).dragTo(block, {
+      force: true,
+      targetPosition: {
+        x: 50,
+        y: 50,
+      },
+    });
+  }
+
+  async prependNewBlockTo(opcode: string, block: Locator) {
+    await this.scrollBlockIntoView(block);
+
+    await this.getBlockInToolbox(opcode).dragTo(block, {
+      force: true,
+      targetPosition: {
+        x: 50,
+        y: 0,
+      },
+    });
+  }
+
+  async countBlocksInStack(stack: Locator) {
+    return stack.evaluate(
+      (el) =>
+        (el
+          // querySelector only looks at the subtree but we also want to match the element itself
+          ?.closest("g.blocklyDraggable")
+          ?.querySelectorAll(".blocklyDraggable").length ?? -1) +
+        // therefore, we add 1 to account for the block matched by the closest call
+        // since querySelectorAll matches only the children
+        1,
+    );
+  }
+
+  async countBlocksInParentStack(stack: Locator) {
+    return stack.evaluate(
+      (el) =>
+        (el
+          ?.closest("g.blocklyDraggable:not(.blocklyDraggable g)")
+          // again, because querySelector only looks at the subtree, we need to go back up another level
+          ?.querySelectorAll(".blocklyDraggable").length ?? -1) + 1,
+    );
```

Comment:
I suspect prettier is messing up the formatting here, but I'm a bit confused by the comments.

`querySelectorAll` returns a `NodeList`, why do we want to add `1` to that? 🤔

Reply:
I tried to make it more readible, what do you think?